File: packages/velix_crypto/lib/src/sender_keys.dart
import 'dart:typed_data';

import 'exceptions.dart';
import 'types.dart';

/// Group session for Sender Keys messaging.
class GroupSession {
  GroupSession._(this._handle, this.groupId);

  final int _handle;
  final Uint8List groupId; // 16 bytes

  static GroupSession create({required Uint8List groupId, required IdentityPublicKey senderId}) {
    throw const VelixCryptoException(
      CryptoErrorCode.protocolError,
      'cryptocore sender_keys FFI not yet wired (Sprint 1)',
    );
  }

  /// Produce the distribution message to send to peers.
  Uint8List distributionMessage() {
    throw const VelixCryptoException(
      CryptoErrorCode.protocolError,
      'cryptocore sender_keys FFI not yet wired (Sprint 1)',
    );
  }

  /// Process a peer's distribution message into the local store.
  void processDistribution(Uint8List distribution) {
    throw const VelixCryptoException(
      CryptoErrorCode.protocolError,
      'cryptocore sender_keys FFI not yet wired (Sprint 1)',
    );
  }

  Ciphertext encrypt(Uint8List plaintext) {
    throw const VelixCryptoException(
      CryptoErrorCode.protocolError,
      'cryptocore sender_keys FFI not yet wired (Sprint 1)',
    );
  }

  Uint8List decrypt(Ciphertext ciphertext) {
    throw const VelixCryptoException(
      CryptoErrorCode.decryptFailed,
      'cryptocore sender_keys FFI not yet wired (Sprint 1)',
    );
  }

  void dispose() {}
}

File: cryptocore/src/sender_keys.rs
//! Sender Keys for group messaging.
//!
//! Wraps libsignal's GroupSessionBuilder + GroupCipher.
//!
//! See `docs/phase-7/07-sender-keys.md`.

use crate::error::{CryptoError, CryptoResult};

/// Distinguishes a (sender, group) tuple in the local store.
pub struct SenderKeyName {
    pub group_id: [u8; 16],
    pub sender_id: [u8; 32],
}

/// Generate a fresh sender-key distribution message for a new group session.
///
/// Final implementation: libsignal `GroupSessionBuilder::create`.
pub fn create_distribution(_name: &SenderKeyName) -> CryptoResult<Vec<u8>> {
    Err(CryptoError::ProtocolError)
}

/// Process a peer's sender-key distribution message.
///
/// Final implementation: libsignal `GroupSessionBuilder::process`.
pub fn process_distribution(
    _name: &SenderKeyName,
    _distribution_message: &[u8],
) -> CryptoResult<()> {
    Err(CryptoError::ProtocolError)
}

/// Encrypt a plaintext for the group, addressed at the sender's chain.
///
/// Final implementation: libsignal `GroupCipher::encrypt`.
pub fn encrypt(_name: &SenderKeyName, _plaintext: &[u8]) -> CryptoResult<Vec<u8>> {
    Err(CryptoError::ProtocolError)
}

/// Decrypt a sender-key ciphertext.
///
/// Final implementation: libsignal `GroupCipher::decrypt`.
pub fn decrypt(_name: &SenderKeyName, _ciphertext: &[u8]) -> CryptoResult<Vec<u8>> {
    Err(CryptoError::DecryptFailed)
}
